File: 387.字符串中的第一个唯一字符.java
/*
 * @lc app=leetcode.cn id=387 lang=java
 *
 * [387] 字符串中的第一个唯一字符
 */

// @lc code=start
import java.util.LinkedHashMap;
class Solution {
    public int firstUniqChar(String s) {
        // 我的
        // if (s.length() == 0)
        //     return -1;
        // int[] idx = new int[26];
        // int j = 0;
        // for (int i = 0; i < 26; i++)
        //     idx[i] = -1;
        // for (int i = 0; i < s.length(); i++) {
        //     j = s.charAt(i)-97;
        //     if (idx[j] == -1)
        //         idx[j] = i;
        //     else if (idx[j] == -2)
        //         continue;
        //     else
        //         idx[j] = -2;
        // }
        // int min = s.length();
        // for (Integer i: idx) {
        //     if (i >= 0 && i < min)
        //         min = i;
        // }
        // return min == s.length() ? -1: min;

        LinkedHashMap<Character, Integer> count = new LinkedHashMap<Character, Integer>();
        int n = s.length();
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            count.put(c, count.getOrDefault(c, 0) + 1);
        }
        for (int i = 0; i < n; i++) {
            if (count.get(s.charAt(i)) == 1) 
                return i;
        }
        return -1;
    }
}
// @lc code=end

File: 387.字符串中的第一个唯一字符.py
#
# @lc app=leetcode.cn id=387 lang=python3
#
# [387] 字符串中的第一个唯一字符
#

# @lc code=start
from collections import Counter
# from collections import defaultdict
# from collections import OrderedDict


class Solution:
    def firstUniqChar(self, s: str) -> int:
        # 1
        # counter = Counter(s)
        # m = len(s) + 1
        # if m == 1:
        #     return -1
        # min_ch = s[-1]
        # for ch in counter:
        #     if counter[ch] == 1:
        #         min_ch = ch
        #         m = min(m, s.index(min_ch))
        # return -1 if m == len(s) + 1 else m
        # 2
        # d = OrderedDict()
        # for ch in s:
        #     if ch not in d:
        #         d[ch] = 0
        #     d[ch] += 1
        # for ch in d:
        #     if d[ch] == 1:
        #         return s.find(ch)
        # return -1
        # 3
        counter = Counter(s)
        for i, ch in enumerate(s):
            if counter[ch] == 1:
                return i
        return -1
# @lc code=end
